--- Day 7: The Treachery of Whales ---

A giant whale has decided your submarine is its next meal, and it's much faster than you are. There's nowhere to run!

Suddenly, a swarm of crabs (each in its own tiny submarine - it's too deep for them otherwise) zooms in to rescue you! They seem to be preparing to blast a hole in the ocean floor; sensors indicate a massive underground cave system just beyond where they're aiming!

The crab submarines all need to be aligned before they'll have enough power to blast a large enough hole for your submarine to get through. However, it doesn't look like they'll be aligned before the whale catches you! Maybe you can help?

There's one major catch - crab submarines can only move horizontally.

You quickly make a list of the horizontal position of each crab (your puzzle input). Crab submarines have limited fuel, so you need to find a way to make all of their horizontal positions match while requiring them to spend as little fuel as possible.

For example, consider the following horizontal positions:

```
16,1,2,0,4,2,7,1,2,14
```

This means there's a crab with horizontal position 16, a crab with horizontal position 1, and so on.

Each change of 1 step in horizontal position of a single crab costs 1 fuel. You could choose any horizontal position to align them all on, but the one that costs the least fuel is horizontal position 2:

```
Move from 16 to 2: 14 fuel
Move from 1 to 2: 1 fuel
Move from 2 to 2: 0 fuel
Move from 0 to 2: 2 fuel
Move from 4 to 2: 2 fuel
Move from 2 to 2: 0 fuel
Move from 7 to 2: 5 fuel
Move from 1 to 2: 1 fuel
Move from 2 to 2: 0 fuel
Move from 14 to 2: 12 fuel
```

This costs a total of 37 fuel. This is the cheapest possible outcome; more expensive outcomes include aligning at position 1 (41 fuel), position 3 (39 fuel), or position 10 (71 fuel).

Determine the horizontal position that the crabs can align to using the least fuel possible. How much fuel must they spend to align to that position?

In [37]:
inp = list(map(int, """
16,1,2,0,4,2,7,1,2,14
""".strip().split(',')))

In [40]:
inp = list(map(int, """
1101,1,29,67,1102,0,1,65,1008,65,35,66,1005,66,28,1,67,65,20,4,0,1001,65,1,65,1106,0,8,99,35,67,101,99,105,32,110,39,101,115,116,32,112,97,115,32,117,110,101,32,105,110,116,99,111,100,101,32,112,114,111,103,114,97,109,10,11,39,71,27,749,252,590,80,445,1125,1044,690,307,145,713,709,468,135,24,330,932,110,1732,1550,826,1,188,1585,388,977,205,351,550,1,782,111,225,523,325,188,134,954,139,448,903,1225,161,791,1539,918,113,15,234,377,2,22,52,1530,119,922,658,41,256,192,161,102,109,8,1492,765,1277,92,245,73,281,1579,485,645,962,287,207,609,109,203,1181,867,371,290,1582,201,16,298,626,329,756,466,433,384,871,84,1661,1177,28,169,263,527,18,619,823,961,41,211,142,842,175,421,202,440,201,104,814,1059,80,17,420,749,522,261,80,540,36,708,1166,357,639,45,224,852,1209,1400,303,13,255,959,83,85,339,184,616,192,688,1145,200,26,376,624,165,10,492,11,866,531,28,20,614,3,500,354,650,74,1156,743,350,8,1641,38,254,673,523,242,186,756,98,1352,425,165,1072,1072,441,487,489,408,685,1083,527,172,32,419,699,672,464,1986,312,941,991,639,260,70,4,490,386,1079,883,285,260,91,204,1724,555,547,468,140,372,94,6,1189,165,439,22,261,255,197,81,9,306,23,61,151,477,326,1393,69,1246,191,113,172,806,116,644,23,500,567,276,237,257,482,262,87,561,530,786,312,1130,251,1391,0,478,85,129,714,64,499,484,771,461,876,563,587,20,4,704,84,230,667,1099,187,491,25,382,1607,14,434,902,1424,263,58,13,394,21,199,885,1113,105,238,423,972,1270,131,103,21,1125,3,1032,122,111,551,122,127,103,1164,1038,688,98,468,547,464,766,1044,1758,718,1476,1057,49,1526,118,391,629,241,280,303,0,915,88,656,1113,100,458,364,1612,362,1090,391,112,722,538,38,54,218,279,893,128,947,336,100,99,1329,342,543,1061,626,40,164,639,640,525,1121,714,592,696,260,103,520,160,151,462,260,1153,1319,132,608,1469,261,1080,96,908,383,36,70,1082,393,99,1573,567,19,1238,391,39,58,334,732,193,79,124,538,1511,185,608,538,0,52,13,190,180,49,184,143,655,74,315,105,906,569,1037,110,224,68,840,1300,130,762,269,560,655,156,819,1137,1483,1037,238,854,91,329,25,677,317,58,563,39,456,919,485,512,37,217,281,1020,914,289,1196,1060,440,1402,85,1130,931,226,325,598,482,861,60,610,308,551,178,1121,436,155,89,1283,887,163,516,1011,26,493,371,538,688,11,1023,148,174,150,401,1115,111,500,84,585,1,307,863,11,1333,414,855,340,810,75,79,292,8,1296,50,17,638,941,260,1250,817,311,184,902,623,97,57,75,425,647,33,225,85,335,34,905,44,644,208,386,194,1427,204,9,226,1278,165,776,510,707,541,378,26,265,661,475,803,344,107,947,307,453,1305,1589,999,55,1364,130,745,545,650,1223,783,965,416,737,275,32,166,951,30,947,1208,159,257,648,214,94,710,386,923,772,910,688,322,312,818,361,77,1,828,280,582,153,428,42,68,1061,271,627,578,12,449,14,583,128,925,355,608,34,492,231,101,6,732,791,94,305,975,1023,201,184,580,591,1604,547,528,417,684,768,1095,925,1429,417,1014,367,1031,160,820,206,227,1045,828,8,170,235,222,483,479,106,182,121,231,282,646,1165,21,68,51,130,14,68,330,1087,76,782,593,85,203,949,164,21,73,14,1590,137,436,163,582,725,46,326,436,59,44,621,317,269,216,486,396,194,713,4,1072,327,186,225,701,320,726,1220,331,17,78,558,1213,1000,436,240,940,1658,19,553,1498,660,152,304,267,33,926,340,556,1098,1434,1813,111,245,1,132,224,782,209,1142,1412,40,716,1,91,184,218,3,484,252,71,263,468,624,12,924,147,687,330,342,746,607,17,709,16,431,232,444,9,1386,44,181,1128,876,74,637,549,213,476,18,320,387,106,61,319,785,165,735,73,806,718,188,327,453,1555,978,13,400,1025,670,698,610,171,97,644,963,1037,53,2,106,20,855,17,7,495,1508,41,205,467,1848,59,243,758,6,396,22,1558,198,227,495,1370,4,128,371,573,1130,591,190,118,1246,1049,122,15,1755,332,567,202,162,989,1061,232,591,308,1433,132,100,564,436,466,15,1229,280,49,1827,701,118,106,246,18,2,184,43,1496,1091,156,334,730,354,429,6,83,1185,1,229,229,16,207,419,1018,197,1492,1805,1334,246,1378,175,297,1362,470,599,263,656,390,119,692,22,859,22,60,1009,5,1310,161,461,157,247,309,195,5,663,402
""".strip().split(',')))

In [2]:
inp

[16, 1, 2, 0, 4, 2, 7, 1, 2, 14]

In [3]:
# part 1

In [16]:
def solve1(data):
    pos_costs = [0 for _ in data]
    
    too_much = 99999999999999
    cur_min = too_much
    
    for idx, align_pos in enumerate(data):
#         print(f'aligning on {align_pos}')
        cost = 0
        for crab_pos in data:
            cost += abs(crab_pos - align_pos)
#             print('cost', cost)
            if cost > cur_min:
                pos_costs[idx] = too_much
                break
        else:
            pos_costs[idx] = cost
            
        cur_min = min(cur_min, cost)
        
    return min(pos_costs), pos_costs
            
solve1(inp)    

(340052,
 [690944,
  465516,
  440898,
  413144,
  99999999999999,
  99999999999999,
  99999999999999,
  99999999999999,
  99999999999999,
  99999999999999,
  99999999999999,
  99999999999999,
  99999999999999,
  99999999999999,
  99999999999999,
  99999999999999,
  413144,
  99999999999999,
  99999999999999,
  99999999999999,
  99999999999999,
  99999999999999,
  99999999999999,
  99999999999999,
  99999999999999,
  99999999999999,
  99999999999999,
  99999999999999,
  393610,
  99999999999999,
  99999999999999,
  392518,
  99999999999999,
  390414,
  99999999999999,
  387898,
  99999999999999,
  99999999999999,
  385546,
  385090,
  99999999999999,
  99999999999999,
  99999999999999,
  99999999999999,
  99999999999999,
  384640,
  99999999999999,
  99999999999999,
  99999999999999,
  99999999999999,
  99999999999999,
  99999999999999,
  99999999999999,
  99999999999999,
  99999999999999,
  99999999999999,
  99999999999999,
  99999999999999,
  99999999999999,
  99999999999999,
  99999

Your puzzle answer was `340052`.

The first half of this puzzle is complete! It provides one gold star: *

--- Part Two ---

The crabs don't seem interested in your proposed solution. Perhaps you misunderstand crab engineering?

As it turns out, crab submarine engines don't burn fuel at a constant rate. Instead, each change of 1 step in horizontal position costs 1 more unit of fuel than the last: the first step costs 1, the second step costs 2, the third step costs 3, and so on.

As each crab moves, moving further becomes more expensive. This changes the best horizontal position to align them all on; in the example above, this becomes 5:

```
Move from 16 to 5: 66 fuel
Move from 1 to 5: 10 fuel
Move from 2 to 5: 6 fuel
Move from 0 to 5: 15 fuel
Move from 4 to 5: 1 fuel
Move from 2 to 5: 6 fuel
Move from 7 to 5: 3 fuel
Move from 1 to 5: 10 fuel
Move from 2 to 5: 6 fuel
Move from 14 to 5: 45 fuel
```

This costs a total of 168 fuel. This is the new cheapest possible outcome; the old alignment position (2) now costs 206 fuel instead.

Determine the horizontal position that the crabs can align to using the least fuel possible so they can make you an escape route! How much fuel must they spend to align to that position?

In [41]:
_cost_cache = {}

def compute_cost(n):
    # https://en.wikipedia.org/wiki/Triangular_number
    if n in _cost_cache:
        return _cost_cache[n]
        
    res = 0
    for i in range(n):
        res += i+1
        
    _cost_cache[n] = res
    return res
    
    

def solve2(data):
    min_pos = min(data)
    pos_costs = [0 for _ in range(min_pos, max(data) + 1)]
    
    too_much = 99999999999999
    cur_min = too_much
    
    for idx, _ in enumerate(pos_costs):
        align_pos = idx + min_pos
#         print(f'aligning on {align_pos}')
        cost = 0
        for crab_pos in data:
            cost += compute_cost(abs(crab_pos - align_pos))
#             print('cost', cost)
            if cost > cur_min:
                pos_costs[idx] = too_much
                break
        else:
            pos_costs[idx] = cost
            
        cur_min = min(cur_min, cost)
        
    return min(pos_costs), pos_costs
            
solve2(inp)    
# compute_cost(16)

(92948968,
 [201818675,
  201352177,
  200886690,
  200422206,
  199958725,
  199496249,
  199034775,
  198574305,
  198114836,
  197656372,
  197198911,
  196742452,
  196286997,
  195832544,
  195379095,
  194926650,
  194475208,
  194024769,
  193575335,
  193126904,
  192679475,
  192233050,
  191787629,
  191343213,
  190899799,
  190457386,
  190015975,
  189575567,
  189136160,
  188697756,
  188260353,
  187823951,
  187388549,
  186954154,
  186520761,
  186088370,
  185656981,
  185226594,
  184797208,
  184368824,
  183941444,
  183515066,
  183089691,
  182665317,
  182241944,
  181819574,
  181398205,
  180977837,
  180558469,
  180140101,
  179722736,
  179306372,
  178891009,
  178476648,
  178063288,
  177650929,
  177239571,
  176829213,
  176419856,
  176011502,
  175604150,
  175197800,
  174792452,
  174388104,
  173984756,
  173582409,
  173181067,
  172780727,
  172381390,
  171983057,
  171585725,
  171189395,
  170794067,
  170399739,
  170006414,
  169614092,
 

In [28]:
compute_cost(16-5)

66